### PR TITLE
Fix: ban  notify from contract developer

### PR DESCRIPTION
**Pull request by Jim8y**
Closes https://github.com/neo-project/neo-devpack-dotnet/issues/843

### CODE REVIEW

**Comment by Jim8y** [comment]
@shargon @cschuchardt88 @roman-khimov @vncoelho @superboyiii please check

**Comment by roman-khimov** [comment]
You can't write any meaningful contract with this PR merged. I'm not sure what's the problem really, just have a proper manifest (_how_ you create it is a separate matter) for your contract and that's it.

**Comment by Jim8y** [comment]
> You can't write any meaningful contract with this PR merged. I'm not sure what's the problem really, just have a proper manifest (_how_ you create it is a separate matter) for your contract and that's it.

What problem are you specifying? user can still call Notify via event, existing contract will still work.

**Comment by shargon** [comment]
> > You can't write any meaningful contract with this PR merged. I'm not sure what's the problem really, just have a proper manifest (_how_ you create it is a separate matter) for your contract and that's it.
> 
> What problem are you specifying? user can still call Notify via event, existing contract will still work. The only thing that I banned is directly calling Notify without defining an event...... cause it will fail the execution for sure anyway.
> 
> Well, this actually only affects the C# compiler,,,, the core remains the same.

But is not possible to call a dynamic event

**Comment by Jim8y** [comment]
> But is not possible to call a dynamic event

what you mean dynamic event? can you please provide an example?

**Comment by roman-khimov** [comment]
@Jim8y, maybe I just don't know enough about C# devpack, if `event` wrapper allows to emit notifications without direct `Runtime.Notify()` calls in the code, then it's not _that_ bad. Still somewhat limiting to me. I can have a pair of create/delete events with the same parameters and emitting them with `Runtime.Notify(nameVariable, params)` might be convenient.

**Comment by Jim8y** [comment]
> I can have a pair of create/delete events with the same parameters and emitting them with `Runtime.Notify(nameVariable, params)` might be convenient.

But current `Notify` checking logic has banned from using arbitary `nameVariable`.  you must defined them as 

```C#
    public delegate void OnSayHelloDelegate(string message);

    [DisplayName("SayHello1")]
    public static event OnSayHelloDelegate OnSayHello;

    [DisplayName("SayHello2")]
    public static event Action<string> SayHello;
```

but if you already defined them this way, using `Notify` would be unecessary, yet keeping it makes it misunderstanding as it can take arbitary name without syntax error while actually arbitary name will fail the execution.

**Comment by shargon** [comment]
> > But is not possible to call a dynamic event
> 
> Current core only accept events that are defined as `event`, what you mean `dynamic event`, i have never heard of this terminology in neo? can you please provide an example?

I mean that maybe you want to emit a event with a dynamic name, who knows..

I think we should not remove notify because if you want to call it by syscall you will be able to do it, so in reality you are not banning it, you are making it difficult

**Comment by cschuchardt88** [comment]
@shargon 
make sense, but it's very misleading when it compiles when it doesn't check contract manifest of parameters for validation. Putting any name in the input of `Runtime.Notify` without the name in manifest will error at execution of the invoke of the contract.

**Comment by shargon** [comment]
> @shargon Makes sense, but it's very misleading when it compiles when it doesn't check contract manifest or parameters for validation. Putting any name in the input of `Runtime.Notify` without the name in manifest will error at execution of the invoke of the contract.

But currently the core allow to call notify with a event that is not present in the manifest....

**Comment by cschuchardt88** [comment]
And old contracts will still work fine. You'll just have to do it with `opcode` at low level. But at high level it should be removed.

**Comment by shargon** [comment]
> Since we don't support events when they don't exist in manifest.

We do

**Comment by cschuchardt88** [comment]
Do you at remember last update or so we remove this functionally. Because of NEP-14

`Inner InvalidOperationException: Event 'SayHello3' does not exist.`

**Comment by cschuchardt88** [comment]
@shargon see https://github.com/neo-project/neo/pull/2810

**Comment by Jim8y** [comment]
> I mean that maybe you want to emit a event with a dynamic name, who knows..
> 
> I think we should not remove notify because if you want to call it by syscall you will be able to do it, so in reality you are not banning it, you are making it difficult

It is not an issue of allowing user to call it or not, if i did not get you wrong, your so called `dynamic notify` is already banned by Erik in the core, just check the code. Yes, you can make it easier to use Notify, but that will not work at all, it will just fail the execution by all means. 

**`Dynamic Notify/Event` is BANNED BY THE CORE**, which is proposed by roman, and implemented by erik.

**Comment by Jim8y** [comment]
> > Since we don't support events when they don't exist in manifest.
> 
> We do

We dont, new logic since the hardfork.

**Comment by shargon** [comment]
> > > Since we don't support events when they don't exist in manifest.
> > 
> > 
> > We do
> 
> @shargon We dont, new logic since the hardfork. Just check the issue and code please https://github.com/neo-project/neo/blob/c78ac5adbd790b1fc60d3ea9abc87cf3eee74a6e/src/Neo/SmartContract/ApplicationEngine.Runtime.cs#L352

Yes, you are right...

File: src/Neo.SmartContract.Framework/Services/Runtime.cs [hunk 1]
```diff
@@ -95,8 +95,8 @@ public static class Runtime
         [Syscall("System.Runtime.Log")]
         public static extern void Log(string message);
 
-        [Syscall("System.Runtime.Notify")]
-        public static extern void Notify(string eventName, object[] state);
+        // [Syscall("System.Runtime.Notify")]
+        // public static extern void Notify(string eventName, object[] state);
```
Comment (shargon) [inline src/Neo.SmartContract.Framework/Services/Runtime.cs:99]:
```suggestion
        [Obsolete]
        [Syscall("System.Runtime.Notify")]
        public static extern void Notify(string eventName, object[] state);
```

**Comment by roman-khimov** [comment]
> Dynamic Notify/Event is BANNED BY THE CORE,

Depends on your definition of "dynamic". If it's "I want to emit anything regardless of manifest" then yes. How is it related to `Runtime.Notify` though? It's just an interop call, it doesn't care. Banning an interop from devpack solves nothing to me. In most cases it shouldn't be used, but I think I still can write some code with it that will work (make event definitions for manifest, but emit via `Runtime.Notify`).

**Comment by cschuchardt88** [comment]
@roman-khimov
Yes you are correct. But It isn't intended to be used in this new way. It was for dynamic events **only** _(all the logic for it in nccs)_. Also alot of contract developers use `Runtime.Notify`, and I think its do to not knowing how to program in `C#` or the outdated document on `neo.org` or even it was easier to emit events then adding an extra 3-4 lines.

**Comment by Jim8y** [comment]
@roman-khimov truth is even shargon and i did not know this, are you expecting developers to use it correctly? how would you feel to use an interface that accepts arbitrary input without emiting any errors during compile but actually will fail the execution without saying why? you love that?

**Comment by cschuchardt88** [comment]
>can not find assistence from anywhere?

Maybe should we should add `xml` code comments to `nccs`. I don't why there isn't any to begin with.

**Comment by Jim8y** [comment]
> > can not find assistence from anywhere?
> 
> Maybe should we should add `xml` code comments to `nccs`. I don't why there isn't any to begin with.

Very general question is, how would developer know they can not do this if they don't already know they can not do this?

File: src/Neo.SmartContract.Framework/Services/Runtime.cs (hunk 1 above)
Comment (Jim8y) [reply src/Neo.SmartContract.Framework/Services/Runtime.cs:99]: Obsolete? it is not Obsoleted,,,,better to do in the analyzer as I have updated in https://github.com/neo-project/neo-devpack-dotnet/pull/839

**Comment by roman-khimov** [comment]
@cschuchardt88, @Jim8y, I see it in a different way. Under the hood it's all `Runtime.Notify` anyway. There is no way to escape calling it in the resulting bytecode. It's just a lower layer of abstraction than `event` magic provided by devpack (that does _both_ `Runtime.Notify` and manifest adjustment).

Let's try to approach it from a different angle, how can I make a `Runtime.Notify` invocation in my code without this wrapper and without a proper `event`?

**Comment by Jim8y** [comment]
> @cschuchardt88, @Jim8y, I see it in a different way. Under the hood it's all `Runtime.Notify` anyway. There is no way to escape calling it in the resulting bytecode. It's just a lower layer of abstraction than `event` magic provided by devpack (that does _both_ `Runtime.Notify` and manifest adjustment).
> 
> Let's try to approach it from a different angle, how can I make a `Runtime.Notify` invocation in my code without this wrapper and without a proper `event`?

Have some mercy to the neo contract developer, dont make it even harder to write neo contract. make every interface clear and easy to use please.

**Comment by cschuchardt88** [comment]
>register event to the manifest with whatever names developer pass in

This won't work at `runtime`, If the following is done:
```csharp
public static void HelloWorld(string eventName, string name)
{
    Runtime.Notify(eventName, new[] { $"Hello, {name}" });
}
```

**Comment by Jim8y** [comment]
> > register event to the manifest with whatever names developer pass in
> 
> This won't work at `runtime`, If the following is done:
> 
> ```cs
> public static void HelloWorld(string eventName, string name)
> {
>     Runtime.Notify(eventName, new[] { $"Hello, {name}" });
> }
> ```
> 
> Which leaves us at the same problem to begin with.

Have you tried it in that pr? This syntax will not pass the compile as it is not a constant string.

**Comment by cschuchardt88** [comment]
>Have you tried it in that pr? This syntax will not pass the compile as it is not a constant string. When I say whatever the name, the name must be known at compile time.

Which PR? but i think _below_ will give you `const`
```csharp
Runtime.Notify($"{eventName}", new[] { $"Hello, {name}" });
```

**Comment by Jim8y** [comment]
https://github.com/neo-project/neo-devpack-dotnet/pull/846 This one

**Comment by vncoelho** [comment]
I am fine with banning it, in fact its directly use has been a burden after events.

**Comment by vncoelho** [comment]
I am fine with banning it, in fact its directly use has been a burden after events.
Banning it can be like this proposal, just remove from the devpack.

File: src/Neo.SmartContract.Framework/Services/Runtime.cs [hunk 2]
```diff
@@ -95,8 +95,8 @@ public static class Runtime
         [Syscall("System.Runtime.Log")]
         public static extern void Log(string message);
 
-        [Syscall("System.Runtime.Notify")]
-        public static extern void Notify(string eventName, object[] state);
+        // [Syscall("System.Runtime.Notify")]
```
Comment (shargon) [inline src/Neo.SmartContract.Framework/Services/Runtime.cs:98]:
```suggestion
        // Events not present in the ABI were disabled in HF_Basilisk
        // [Syscall("System.Runtime.Notify")]
```

**Review by shargon — APPROVED** [review]

**Pull request merged**